File: 06_Assembler/assembler.py
import sys
import os

dest_table = {
    "null": "000", "M": "001", "D": "010", "MD": "011",
    "A": "100", "AM": "101", "AD": "110", "AMD": "111"
}

jump_table = {
    "null": "000", "JGT": "001", "JEQ": "010", "JGE": "011",
    "JLT": "100", "JNE": "101", "JLE": "110", "JMP": "111"
}

comp_table = {
    "0":   "0101010", "1":   "0111111", "-1":  "0111010",
    "D":   "0001100", "A":   "0110000", "M":   "1110000",
    "!D":  "0001101", "!A":  "0110001", "!M":  "1110001",
    "-D":  "0001111", "-A":  "0110011", "-M":  "1110011",
    "D+1": "0011111", "A+1": "0110111", "M+1": "1110111",
    "D-1": "0001110", "A-1": "0110010", "M-1": "1110010",
    "D+A": "0000010", "D+M": "1000010", "D-A": "0010011",
    "D-M": "1010011", "A-D": "0000111", "M-D": "1000111",
    "D&A": "0000000", "D&M": "1000000", "D|A": "0010101", "D|M": "1010101"
}

symbol_table = {
    "SP": 0, "LCL": 1, "ARG": 2, "THIS": 3, "THAT": 4,
    "R0": 0, "R1": 1, "R2": 2, "R3": 3, "R4": 4, "R5": 5, "R6": 6, "R7": 7,
    "R8": 8, "R9": 9, "R10": 10, "R11": 11, "R12": 12, "R13": 13, "R14": 14, "R15": 15,
    "SCREEN": 16384, "KBD": 24576
}

file = sys.argv[1]
output_lines = []
rom_address = 0
instructions_only = []
next_variable_address = 16
base_name = os.path.splitext(file)[0]
output_filename = base_name + ".hack"

with (open(file, "r") as f):
    for line in f:
        clean_line = line.split("//")[0].strip()
        if not clean_line:
            continue

        if clean_line.startswith("(") and clean_line.endswith(")"):
            label = clean_line[1:-1].strip()
            symbol_table[label] = rom_address
        else:
            instructions_only.append(clean_line)
            rom_address += 1

for line in instructions_only:
    if line.startswith("@"):
        symbol = line[1:].strip()

        if symbol.isdigit():
            value = int(symbol)
        elif symbol in symbol_table:
            value = symbol_table[symbol]
        else:
            symbol_table[symbol] = next_variable_address
            value = next_variable_address
            next_variable_address += 1

        binary_a = format(value, "016b")
        output_lines.append(binary_a)

    else:
        dest_part = "null"
        jump_part = "null"

        if "=" in line:
            dest_part, rest = line.split("=")
        else:
            rest = line

        if ";" in rest:
            comp_part, jump_part = rest.split(";")
        else:
            comp_part = rest

        c_bits = comp_table[comp_part]
        d_bits = dest_table[dest_part]
        j_bits = jump_table[jump_part]

        binary_c = "111" + c_bits + d_bits + j_bits
        output_lines.append(binary_c)

with open(output_filename, "w") as f:
    f.write("\n".join(output_lines))

File: 06_Assembler/assembler.java
import java.io.*;
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;

class Assembler {
    public static void main(String[] args) {
        Map<String, String> destTable = Map.of(
                "null", "000", "M", "001", "D", "010", "MD", "011",
                "A", "100", "AM", "101", "AD", "110", "AMD", "111"
        );

        Map<String, String> jumpTable = Map.of(
                "null", "000", "JGT", "001", "JEQ", "010", "JGE", "011",
                "JLT", "100", "JNE", "101", "JLE", "110", "JMP", "111"
        );

        Map<String, String> compTable = new HashMap<>();
        Map<String, Integer> symbolTable = new HashMap<>();

        compTable.put("0",   "0101010"); compTable.put("1",   "0111111");
        compTable.put("-1",  "0111010"); compTable.put("D",   "0001100");
        compTable.put("A",   "0110000"); compTable.put("M",   "1110000");
        compTable.put("!D",  "0001101"); compTable.put("!A",  "0110001");
        compTable.put("!M",  "1110001"); compTable.put("-D",  "0001111");
        compTable.put("-A",  "0110011"); compTable.put("-M",  "1110011");
        compTable.put("D+1", "0011111"); compTable.put("A+1", "0110111");
        compTable.put("M+1", "1110111"); compTable.put("D-1", "0001110");
        compTable.put("A-1", "0110010"); compTable.put("M-1", "1110010");
        compTable.put("D+A", "0000010"); compTable.put("D+M", "1000010");
        compTable.put("D-A", "0010011"); compTable.put("D-M", "1010011");
        compTable.put("A-D", "0000111"); compTable.put("M-D", "1000111");
        compTable.put("D&A", "0000000"); compTable.put("D&M", "1000000");
        compTable.put("D|A", "0010101"); compTable.put("D|M", "1010101");

        symbolTable.put("SP", 0); symbolTable.put("LCL", 1);
        symbolTable.put("ARG", 2); symbolTable.put("THIS", 3);
        symbolTable.put("THAT", 4); symbolTable.put("SCREEN", 16384);
        symbolTable.put("KBD", 24576);
        for (int i = 0; i <= 15; i++) {
            symbolTable.put("R" + i, i);
        }

        int romAddress = 0;
        int nextVariableAddress = 16;
        ArrayList<String> instructionsOnly = new ArrayList<>();
        ArrayList<String> outputLines = new ArrayList<>();
        String inputFileName = args[0];

        try (BufferedReader reader = new BufferedReader(new FileReader(inputFileName))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String cleanLine = line.split("//")[0].trim();
                if (cleanLine.isEmpty()) {
                    continue;
                }

                if (cleanLine.startsWith("(") && cleanLine.endsWith(")")) {
                    String label = cleanLine.substring(1, cleanLine.length() - 1).trim();
                    symbolTable.put(label, romAddress);
                } else {
                    instructionsOnly.add(cleanLine);
                    romAddress++;
                }
            }
        } catch (IOException e) {
            System.err.println("Fehler beim Lesen der Datei: " + e.getMessage());
            return;
        }

        for (String line : instructionsOnly) {
            if (line.startsWith("@")) {
                String symbol = line.substring(1).trim();
                int value;

                if (symbol.matches("\\d+")) {
                    value = Integer.parseInt(symbol);
                } else if (symbolTable.containsKey(symbol)) {
                    value = symbolTable.get(symbol);
                } else {
                    symbolTable.put(symbol, nextVariableAddress);
                    value = nextVariableAddress;
                    nextVariableAddress++;
                }

                String binaryA = String.format("%16s", Integer.toBinaryString(value)).replace(' ', '0');
                outputLines.add(binaryA);

            } else {
                String destPart = "null";
                String jumpPart = "null";
                String compPart;
                String rest = line;

                if (rest.contains("=")) {
                    String[] parts = rest.split("=", 2);
                    destPart = parts[0];
                    rest = parts[1];
                }

                if (rest.contains(";")) {
                    String[] parts = rest.split(";", 2);
                    compPart = parts[0];
                    jumpPart = parts[1];
                } else {
                    compPart = rest;
                }

                String cBits = compTable.get(compPart);
                String dBits = destTable.get(destPart);
                String jBits = jumpTable.get(jumpPart);

                String binaryC = "111" + cBits + dBits + jBits;
                outputLines.add(binaryC);
            }
        }

        String outputFilename;
        int dotIndex = inputFileName.lastIndexOf('.');
        if (dotIndex != -1) {
            outputFilename = inputFileName.substring(0, dotIndex) + ".hack";
        } else {
            outputFilename = inputFileName + ".hack";
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFilename))) {
            for (int i = 0; i < outputLines.size(); i++) {
                writer.write(outputLines.get(i));
                if (i < outputLines.size() - 1) {
                    writer.newLine();
                }
            }
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}